apps/saleor-app-checkout/backend/payments/providers/adyen/utils.ts

```diff
@@ -44,12 +55,209 @@ export const getLineItems = (lines: OrderFragment["lines"]): Types.checkout.Line
   [event.name, event.reference].join();
 
-export const getAmountAfterRefund = (money: MoneyFragment, refundAmount: number) => {
-  const chargedAmount = money.amount - refundAmount;
+export const encodeBasicAuth = (username: string, password: string) => {
+  return Buffer.from(username + ":" + password, "ascii").toString("base64");
+};
+
+export const verifyBasicAuth = (username: string, password: string, token: string) => {
+  const credentials = encodeBasicAuth(username, password);
+  return token === `Basic ${credentials}`;
+};
+
+export const getNotificationStatus = (
+  notification: Types.notification.NotificationRequestItem
+): TransactionStatus => {
+  const { eventCode, success } = notification;
+  if (
+    failedEvents.has(eventCode) ||
+    success === Types.notification.NotificationRequestItem.SuccessEnum.False
+  ) {
+    return "FAILURE";
+  }
+
+  return "SUCCESS";
+};
+
+export type TransactionAmounts = Pick<
+  TransactionUpdateInput,
+  "amountRefunded" | "amountAuthorized" | "amountCharged" | "amountVoided"
+>;
+
+const nonNegative = (amount: number) => (amount < 0 ? 0 : amount);
+
+const getCurrencyFromTransaction = (transaction: TransactionFragment | null) => {
+  if (!transaction) {
+    return undefined;
+  }
+
+  const currencies = [
+    transaction?.refundedAmount?.currency,
+    transaction?.voidedAmount?.currency,
+    transaction?.chargedAmount?.currency,
+    transaction?.authorizedAmount?.currency,
+  ];
+
+  // find first existing currency
+  return currencies.find((item) => typeof item === "string");
+};
+
+export const getTransactionAmountFromAdyen = (
+  notification: Types.notification.NotificationRequestItem,
+  transaction: TransactionFragment | null
+): TransactionAmounts => {
+  const getTransactionAmount = getTransactionAmountGetterAsMoney({
+    voided: transaction?.voidedAmount?.amount ?? 0,
+    charged: transaction?.chargedAmount?.amount ?? 0,
+    refunded: transaction?.refundedAmount?.amount ?? 0,
+    authorized: transaction?.authorizedAmount?.amount ?? 0,
+  });
+  const transactionCurrency = getCurrencyFromTransaction(transaction);
+
+  const notificationAmount = currency(getSaleorAmountFromAdyen(notification.amount.value ?? 0));
+  const notificationCurrency = notification.amount.currency!;
+
+  if (notification.success === Types.notification.NotificationRequestItem.SuccessEnum.False) {
+    return {};
+  }
+
+  if (notification.eventCode === EventCodeEnum.Pending) {
+    return {};
+  }
+
+  if (
+    typeof notification.amount.currency !== "string" ||
+    typeof notification.amount.value !== "number"
+  ) {
+    console.error("(Adyen webhook) Notification without amount or currency");
+    throw new Error("Notification doesn't contain amount or currency");
+  }
 
-  if (chargedAmount < 0) {
-    throw "Amount after refund cannot be negative";
+  if (transactionCurrency && transactionCurrency !== notification.amount.currency) {
```

Comment:
What if there's no `transactionCurrency` ? Shouldn't we fail?

Reply:
No, if a transaction doesn't have currency set, then this means:
- it doesn't have any amount set yet
- there's no transaction, and we need to create it

This should fail only if there **is** a currency, and it doesn't match what came from Adyen. It's unlikely to happen at all, but better safe than sorry 😅